src/client/dfuse/ops/read.c

```diff
@@ -173,186 +173,112 @@ pick_eqt(struct dfuse_info *dfuse_info)
 	int                   rc;
 	int                   entered;
-	ATOMIC int            exited;
-	bool                  exiting;
 	bool                  complete;
+	d_list_t              req_list;
 };
 
-static void
-chunk_free(struct read_chunk_data *cd)
-{
-	d_list_del(&cd->list);
-	d_slab_release(cd->eqt->de_read_slab, cd->ev);
-	D_FREE(cd);
-}
+struct read_chunk_req {
+	d_list_t              req_list;
+	struct dfuse_obj_hdl *oh;
+	fuse_req_t            req;
+	int                   slot;
+};
 
 /* Called when the last open file handle on a inode is closed.  This needs to free everything which
  * is complete and for anything that isn't flag it for deletion in the callback.
  *
  * Returns true if the feature was used.
  */
 bool
-read_chunk_close(struct dfuse_inode_entry *ie)
+read_chunk_close(struct active_inode *active)
 {
 	struct read_chunk_data *cd, *cdn;
-	bool                    rcb = false;
 
-	D_SPIN_LOCK(&ie->ie_active->lock);
-	if (d_list_empty(&ie->ie_active->chunks))
-		goto out;
-
-	rcb = true;
+	if (d_list_empty(&active->chunks))
+		return false;
 
-	d_list_for_each_entry_safe(cd, cdn, &ie->ie_active->chunks, list) {
-		if (cd->complete) {
-			chunk_free(cd);
-		} else {
-			cd->exiting = true;
-		}
+	d_list_for_each_entry_safe(cd, cdn, &active->chunks, list) {
+		D_ASSERT(cd->complete);
+		d_list_del(&cd->list);
+		d_slab_release(cd->eqt->de_read_slab, cd->ev);
+		D_FREE(cd);
 	}
-out:
-	D_SPIN_UNLOCK(&ie->ie_active->lock);
-	return rcb;
+	return true;
 }
 
 static void
 chunk_cb(struct dfuse_event *ev)
 {
 	struct read_chunk_data *cd = ev->de_cd;
 	struct active_inode    *ia = cd->ia;
-	fuse_req_t              req;
-	bool                    done = false;
+	struct read_chunk_req  *cr;
+	struct read_chunk_req  *crn;
 
 	cd->rc = ev->de_ev.ev_error;
 
 	if (cd->rc == 0 && (ev->de_len != CHUNK_SIZE)) {
-		cd->rc = EIO;
 		DS_WARN(cd->rc, "Unexpected short read bucket %ld (%#zx) expected %i got %zi",
 			cd->bucket, cd->bucket * CHUNK_SIZE, CHUNK_SIZE, ev->de_len);
 	}
 
 	daos_event_fini(&ev->de_ev);
 
-	do {
-		int i;
-		req = 0;
+	/* Mark as complete so no more get put on list */
+	D_SPIN_LOCK(&ia->lock);
+	cd->complete = true;
 
-		D_SPIN_LOCK(&ia->lock);
-
-		if (cd->exiting) {
-			chunk_free(cd);
-			D_SPIN_UNLOCK(&ia->lock);
-			return;
-		}
+	/* Mark the slot as replied to.  There's a race here as the slot hasn't been replied to
+	 * however references are dropped by the DFUSE_REPLY macros below so an extra ref on active
+	 * would be required.  The danger is that the bucket gets put on the end of the list rather
+	 * than the start.
+	 */
+	d_list_for_each_entry(cr, &cd->req_list, req_list)
+		cd->slot_done[cr->slot] = true;
 
-		cd->complete = true;
-		for (i = 0; i < 8; i++) {
-			if (cd->reqs[i]) {
-				req         = cd->reqs[i];
-				cd->reqs[i] = 0;
-				break;
-			}
-		}
+	D_SPIN_UNLOCK(&ia->lock);
 
-		D_SPIN_UNLOCK(&ia->lock);
+	d_list_for_each_entry_safe(cr, crn, &cd->req_list, req_list) {
```

Comment:
Note to self: There's no reference on `cd` held here after the last call to `DFUSE_REPLY../.` so the list needs to be spliced onto the stack before the list is iterated.